app/Http/Controllers/admin/StatisticsController.php

```diff
@@ -1,114 +1,215 @@
 
-        $yearlyData = $this->user->whereYear('created_at', $selectedYear)->get();
-
-        // Initialize an array to hold the results
-        $numericalDataNumByAttribute = [];
+    public function showStatistics(){
+        return view('admin.statistics.show');
+    }
 
-        // Loop through each attribute
-        foreach($attributes as $attribute) {
-            // Loop through each month
-            foreach($months as $monthNumber => $monthName) {
-                // Query the database
-                $count = $yearlyData->where('attribute_id', $attribute->id)
-                    ->whereMonth('created_at', $monthNumber)
-                    ->count();
+    public function fetchYearlyStatisticalData(Request $request){
+        $months = self::MONTH_MAP;
+        $selectedYear = $request->selectedYear;
+        $selectedTableId = $request->selectedTableId;
+    
+        $tableIdMethodMap = [
+            'registrations-num' => 
+                [
+                    'yearlyDataFetchMethod' => 'fetchYearlyRegistrationsData', 'table' => 'users', 
+                    'statisticalDataFetchMethod' => 'fetchRegistrationsStatisticalData'
+                ],
+            'deletions-num' => 
+                [
+                    'yearlyDataFetchMethod' => 'fetchYearlyDeletionsData', 
+                    'table' => 'users', 
+                    'statisticalDataFetchMethod' => 'fetchDeletionsStatisticalData'
+                ],
+            'reservations-num' => 
+                [
+                    'yearlyDataFetchMethod' => 'fetchYearlyReservationsData', 'table' => 'reservations', 
+                    'statisticalDataFetchMethod' => 'fetchReservationsStatisticalData'
+                ],
+            'cancellations-num' => 
+                [
+                    'yearlyDataFetchMethod' => 'fetchYearlyCancellationsData', 'table' => 'reservations', 
+                    'statisticalDataFetchMethod' => 'fetchCancellationsStatisticalData'
+                ],
+            'sales-num' => 
+                [
+                    'yearlyDataFetchMethod' => 'fetchYearlySalesData', 
+                    'table' => 'reservations', 
+                    'statisticalDataFetchMethod' => 'fetchSalesStatisticalData'
+                ],
+        ];
+    
+        if (!array_key_exists($selectedTableId, $tableIdMethodMap)) {
+            return response()->json(['error' => 'Invalid table ID']);
+        }
+    
+        $yearlyDataFetchMethod = $tableIdMethodMap[$selectedTableId]['yearlyDataFetchMethod'];
+        $table = $tableIdMethodMap[$selectedTableId]['table'];
+        $statisticalDataFetchMethod = $tableIdMethodMap[$selectedTableId]['statisticalDataFetchMethod'];
 
-                // Store the result in the results array
-                $numericalDataNumByAttribute[$attribute->name][$monthName] = $count;
+        $attributes = $this->getAttributes($table);
+        $yearlyData = $this->$yearlyDataFetchMethod($selectedYear);
```

Comment:
Can you test when $yearlyData is null ? If it won't create a crash / error. Or for safety you can also wrap an if statement ensuring that the $yearlyData is not null / empty

```
if (!empty($yearlyData) {
       $statisticalData = $this->$statisticalDataFetchMethod($attributes, $months, $yearlyData, $selectedTableId);
    
        // Add "Total" attribute
        $statisticalData['Total'] = [];
        foreach($months as $monthName) {
            $total = 0;
            foreach($attributes as $attribute) {
                $total += $statisticalData[$attribute->name][$monthName];
            }
            $statisticalData['Total'][$monthName] = $total;
        }
}
```

Reply:
@Ninit29 
@Tomo-Furu 
Thank you for your feedback.
Let me answer it.

**As for the current version of the codes, I did the following for the test.**

- I replaced `$yearlyData = $this->$yearlyDataFetchMethod($selectedYear);` with `$yearlyData = null;` and the page does not show table and chart as attached.

![image](https://github.com/kredo-school/19tha_easepark/assets/121096971/8d09a66f-d1f9-4766-b23c-0ad870c9cc04)

However, the route for Ajax received the error as expected.
<img width="476" alt="image" src="https://github.com/kredo-school/19tha_easepark/assets/121096971/f0addc37-587a-480d-a4dc-5a2db27fa60b">

Basically, there are not any problems on the UI but with problems in data transaction.

Since you gave us the alternative codes, I modified the codes including the one below.
```
if (!empty($yearlyData)) {
            $statisticalData = $this->$statisticalDataFetchMethod($attributes, $months, $yearlyData, $selectedTableId);
            
            // Add "Total" attribute
            $statisticalData['Total'] = [];
            foreach($months as $monthName) {
                $total = 0;
                foreach($attributes as $attribute) {
                    $total += $statisticalData[$attribute->name][$monthName];
                }
                $statisticalData['Total'][$monthName] = $total;
            }
        }
```

Like the previous version, UI was fine but with some errors for the data transaction.
(Since `$query` does not have any data.)
<img width="999" alt="image" src="https://github.com/kredo-school/19tha_easepark/assets/121096971/4343ce31-8f68-4b2a-81d2-83fb36442003">

Let me know if I should modify the codes.
I have already commit and push the updated version into github.

Reply:
Yes you can keep the if statement, actually the new error is not related to this controller but in the manageStatistics.js, you may want to check if `data.statisticalData[attribute][month] !== 'undefined'` before accessing it. This is just optional for you to do (if just want to have an error less application) since this is a rare scenario. For now, I will merge this PR

`data.months.forEach(function(month) {tableHtml += '<td>' + (data.statisticalData[attribute][month] || '0') + '</td>';`